src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -0,0 +1,251 @@
+        /// timestamp is divisible by 16 (using timestamp mask).
+        /// </summary>
+        /// <param name="header">The proven block header.</param>
+        private void CheckHeaderAndCoinstakeTimes(ProvenBlockHeader header)
+        {
+            uint coinstakeTime = header.Coinstake.Time;
+            uint headerTime = header.Time; 
+
+            // Check if times are equal and coinstake tx time is divisible by 16.
+            if ((headerTime == coinstakeTime) && ((coinstakeTime & PosConsensusOptions.StakeTimestampMask) == 0))
+                return;
+
+            this.Logger.LogTrace("(-)[BAD_TIME]");
+            ConsensusErrors.StakeTimeViolation.Throw();
+        }
+
+        /// <summary>
+        /// Checks the coinstake age requirement.
+        /// </summary>
+        /// <param name="chainedHeader">The chained header.</param>
+        /// <param name="unspentOutputs">The unspent outputs.</param>
+        private void CheckCoinstakeAgeRequirement(ChainedHeader chainedHeader, UnspentOutputs unspentOutputs)
+        {
+            ChainedHeader prevChainedHeader = chainedHeader.Previous;
+
+            var options = (PosConsensusOptions)this.PosParent.Network.Consensus.Options;
+            int targetDepth = options.GetStakeMinConfirmations(prevChainedHeader.Height + 1, this.PosParent.Network) - 1;
+
+            int actualDepth = prevChainedHeader.Height - (int)unspentOutputs.Height;
+            if (actualDepth < targetDepth)
+                return;
+
+            this.Logger.LogTrace("(-)[BAD_STAKE_DEPTH]");
+            ConsensusErrors.InvalidStakeDepth.Throw();
+        }
+
+        /// <see cref="IStakeValidator.VerifySignature"/>
+        private void CheckSignature(ProvenBlockHeader header, UnspentOutputs unspentOutputs)
+        {
+            if (this.stakeValidator.VerifySignature(unspentOutputs, header.Coinstake, 0, ScriptVerify.None))
+                return;
+            
+            this.Logger.LogTrace("(-)[BAD_SIGNATURE]");
+            ConsensusErrors.CoinstakeVerifySignatureFailed.Throw();
+        }
+
+        /// <see cref="IStakeValidator.CheckStakeKernelHash"/>
+        private void CheckStakeKernelHash(PosRuleContext context, UnspentOutputs stakingCoins, ProvenBlockHeader header, ChainedHeader chainedHeader)
+        {
+            OutPoint prevOut = this.GetPreviousOut(header);
+            uint transactionTime = header.Coinstake.Time;
+
+            ChainedHeader prevChainedHeader = chainedHeader.Previous;
+
+            BlockStake prevBlockStake = this.PosParent.StakeChain.Get(prevChainedHeader.HashBlock);
+            if (prevBlockStake == null)
+            {
+                this.Logger.LogTrace("(-)[BAD_PREV_STAKE]");
+                ConsensusErrors.PrevStakeNull.Throw();
+            }
+
+            uint headerBits = chainedHeader.Header.Bits.ToCompact();
+
+            this.stakeValidator.CheckStakeKernelHash(context, headerBits, prevBlockStake, stakingCoins, prevOut, transactionTime);
```

Comment:
Do we not need to check that we are passed the POW era?

Reply:
As per line 58 of the rule we will not even run any validations for this rules if Proven Headers were not activated. So there is no way to reach this validation if POW hasn't passed.